### PR TITLE
chore(caches): use cache module for ReposityCaches/ContextCaches

### PR DESCRIPTION
**Pull request by sileht**
Change-Id: Iadaaafa22d03d9a161604275c94a131f096f455c

### CODE REVIEW

File: mergify_engine/cache.py [hunk 1]
```diff
@@ -59,3 +59,17 @@ def __getitem__(self, key: _K) -> typing.Union[_V, UnsetType]:
...
+    _cache: typing.Union[_V, UnsetType] = dataclasses.field(default=Unset)
+
+    def get(self) -> typing.Union[_V, UnsetType]:
+        return self._cache
+
+    def set(self, value: _V) -> None:
+        self._cache = value
+
+    def delete(self) -> None:
+        self._cache = Unset
```
Comment (jd) [inline mergify_engine/cache.py:75]: What about using the descriptor protocol here?

Comment (sileht) [reply mergify_engine/cache.py:75]: Looks fancy at first glance:

```
assert self._caches.mergify_config is Unset:
self._caches.mergify_config = "value"
assert self._caches.mergify_config == "value"
del self._caches.mergify_config
assert self._caches.mergify_config is Unset
```

Comment (jd) [reply mergify_engine/cache.py:75]: Is this a yes?

Comment (sileht) [reply mergify_engine/cache.py:75]: I just read the doc and it's `ContextCache` that have to be description protocol, but currently the `ContextCache` has `Cache` and `SingleCache` attributes. The value `SingleCache` as-is is the typing is simple and strong. If I have to description the same typing with a descriptor, I have to create an @overload for key the description protocol support.

**Review by jd — APPROVED** [review]

**Pull request merged**